fe/fe-core/src/main/java/com/starrocks/journal/bdbje/BDBJournalCursor.java

```diff
@@ -41,50 +41,123 @@
+    // names of all local databases, will set on initialization, and will update every time `prelong()` is called
+    protected List<Long> localDBNames = null;
+    // index of next db to be opened
+    protected int nextDbPositionIndex = -1;
+    // the database of current log
+    protected CloseSafeDatabase database = null;
 
-    public static BDBJournalCursor getJournalCursor(BDBEnvironment env, long fromKey, long toKey) throws
-            JournalException {
-        if (toKey < fromKey || fromKey < 0) {
-            throw new JournalException(String.format("Invalid key range! fromKey %s toKey %s", fromKey, toKey));
+    /**
+     * handle DatabaseException carefully
+     *
+     * 1. wrap as JournalException and return if it's a normal DatabaseException.
+     * 2. RestartRequiredException is a fatal error since we're replaying as a follower, we must exit by raising an
+     * JournalInconsistentException.
+     * 3. If it is an InsufficientLogException, we should call refreshLog() to restore the data as much as possible.
+     */
+    protected JournalException wrapDatabaseException(DatabaseException originException, String errMsg)
+            throws JournalInconsistentException {
+        if (originException instanceof RestartRequiredException) {
+            errMsg += "Got RestartRequiredException, will exit.";
+            LOG.warn(errMsg, originException);
+            if (originException instanceof InsufficientLogException) {
+                // for InsufficientLogException we should refresh the log and
+                // then exit the process because we may have read dirty data.
+                environment.refreshLog((InsufficientLogException) originException);
+            }
+            JournalInconsistentException journalInconsistentException = new JournalInconsistentException(errMsg);
+            journalInconsistentException.initCause(originException);
+            throw journalInconsistentException;
         }
-        return new BDBJournalCursor(env, fromKey, toKey);
+        LOG.warn(errMsg);
+        JournalException exception = new JournalException(errMsg);
+        exception.initCause(originException);
+        return exception;
     }
 
-    protected BDBJournalCursor(BDBEnvironment env, long fromKey, long toKey) throws JournalException {
-        this.environment = env;
-        this.toKey = toKey;
-        this.currentKey = fromKey;
-        this.dbNames = env.getDatabaseNames();
-        if (dbNames == null) {
-            throw new JournalException("failed to get db names!");
+    /**
+     * init journal cursor
+     * if toKey = -1(CUROSR_END_KEY), it will automatically search the end.
+     */
+    public static BDBJournalCursor getJournalCursor(BDBEnvironment env, long fromKey, long toKey)
+            throws JournalException, JournalInconsistentException, InterruptedException {
+        if (fromKey < 0 || toKey > 0 && toKey < fromKey || toKey <= 0 && toKey != JournalCursor.CUROSR_END_KEY) {
```

Comment:
Could you add some parentheses? Too many conditions

Reply:
Done.